src/ranges.jl

```diff
@@ -0,0 +1,231 @@
+    return unsafe_isempty_one_to(last(r))
+  else
+    return unsafe_isempty_unit_range(first(r), last(r))
+  end
+end
+
+unsafe_isempty_one_to(lst) = lst <= zero(lst)
+unsafe_isempty_unit_range(fst, lst) = fst > lst
+
+unsafe_isempty_unit_range(fst::T, lst::T) where {T} = Integer(lst - fst + one(T))
+
+unsafe_length_one_to(lst::T) where {T<:Int} = T(lst)
+unsafe_length_one_to(lst::T) where {T} = Integer(lst - zero(lst))
+
+Base.@propagate_inbounds function Base.getindex(r::OptionallyStaticUnitRange, i::Integer)
+  if known_first(r) === oneunit(r)
+    return get_index_one_to(r, i)
+  else
+    return get_index_unit_range(r, i)
+  end
+end
+
+@inline function get_index_one_to(r, i)
+  @boundscheck if ((i > 0) & (i <= last(r)))
+    throw(BoundsError(r, i))
+  end
+  return convert(eltype(r), i)
+end
+
+@inline function get_index_unit_range(r, i)
+  val = first(r) + (i - 1)
+  @boundscheck if i > 0 && val <= last(r) && val >= first(r)
+    throw(BoundsError(r, i))
+  end
+  return convert(eltype(r), val)
+end
+
+_try_static(x, y) = Val(x)
+_try_static(::Nothing, y) = Val(y)
+_try_static(x, ::Nothing) = Val(x)
+_try_static(::Nothing, ::Nothing) = nothing
+
+###
+### length
+###
+@inline function known_length(::Type{T}) where {T<:AbstractUnitRange}
+  fst = known_first(T)
+  lst = known_last(T)
+  if fst === nothing || lst === nothing
+    return nothing
+  else
+    if fst === oneunit(eltype(T))
+      return unsafe_length_one_to(lst)
+    else
+      return unsafe_length_unit_range(fst, lst)
+    end
+  end
+end
+
+function Base.length(r::OptionallyStaticUnitRange{T}) where {T}
+  if isempty(r)
+    return zero(T)
+  else
+    if known_one(r) === one(T)
+      return unsafe_length_one_to(last(r))
+    else
+      return unsafe_length_unit_range(first(r), last(r))
+    end
+  end
+end
+
+function unsafe_length_unit_range(fst::T, lst::T) where {T<:Union{Int,Int64,Int128}}
+  return Base.checked_add(Base.checked_sub(lst, fst), one(T))
+end
+function unsafe_length_unit_range(fst::T, lst::T) where {T<:Union{UInt,UInt64,UInt128}}
+  return Base.checked_add(lst - fst, one(T))
+end
+
+"""
+    indices(x[, d])
+
+Given an array `x`, this returns the indices along dimension `d`. If `x` is a tuple
+of arrays then the indices corresponding to dimension `d` of all arrays in `x` are
+returned. If any indices are not equal along dimension `d` an error is thrown. A
+tuple may be used to specify a different dimension for each array. If `d` is not
+specified then indices for visiting each index of `x` is returned.
+"""
+@inline function indices(x)
+  inds = eachindex(x)
+  if inds isa AbstractUnitRange{<:Integer}
+    return Base.Slice(OptionallyStaticUnitRange(inds))
+  else
+    return inds
+  end
+end
+
+function indices(x::Tuple)
+  inds = map(eachindex, x)
+  @assert all(isequal(first(inds)), Base.tail(inds)) "Not all specified axes are equal: $inds"
+  return reduce(_pick_range, inds)
+end
+
+indices(x, d) = indices(axes(x, d))
```

Comment:
This computes the indices of the indices, rather than the indices of `x`. You're implicitly assuming that `axes(x)` is idempotent. The older implementations of OffsetArrays would break this, for example, although the community seems to have widely settled on idempotency as a useful characteristic for the axes.

Reply:
In terms of iterating values it should theoretically be the same, but it's not entirely idempotent because `indices(x::OneTo) = Slice(OptionallyStaticUnitRange(Val(1), last(x)))`. If `indices(x::OffsetArray, d) = indices(axes(x, d))` doesn't produce something that is appropriately offset then it could do something like `indices(x::OffsetArray, d) = indices(range(x.offset[d], stop = size(x, d))`.